### ISSUE TITLE
Strange behavior with Document.set(obj) and arrays (only sometimes)..

### ISSUE THREAD

**Issue by zackurben** [Automattic/mongoose#3315]
Hello, I have been experiencing some strange behavior with Document.set(<object>) and was wondering if I could get some feedback. I have searched for prior issues, and this seems like it could be possibly be related to: https://github.com/Automattic/mongoose/issues/1638

The test I have included does not show my exact use-case but this shows the exact behavior I am experiencing (I am having trouble isolating my exact issue).

Here are my env versions:
```
mongodb: v3.0.3
mongoose: v4.1.3
node: v0.12.6
```

The following test has 3 variations (seen in test code):
```javascript
case 1) var update = result; // works fine
case 2) var update = result.toObject(); // works fine
case 3) var update = _.clone(result._doc); // works fine
case 4) var update = result._doc; // broken, mangles permissions array
```

Here is my test:
```javascript
'use strict';

var mongoose = require('mongoose');
var _ = require('lodash');
mongoose.connect('mongodb://localhost/testvalidation');

var echo = function(value) {
  console.log('\nPermissions:');
  console.log(value);
  return value;
};

var PermissionSchema = {
  _id: false,
  type: {
    type: String
  },
  roles: {
    type: [mongoose.Schema.Types.ObjectId],
    ref: 'role'
  }
};
var RoleSchema = new mongoose.Schema({ name: String });
var DocumentSchema = new mongoose.Schema({
  permissions: {
    type: [PermissionSchema],
    set: echo
  }
});

var Role = mongoose.model('role', RoleSchema);
var Document = mongoose.model('document', DocumentSchema);

var r1 = new Role({ name: 'Admin' });
var r2 = new Role({ name: 'User' });
var d1 = new Document({
  permissions: [
    {type: 'create', roles: [r1._id, r2._id]},
    {type: 'read', roles: [r1._id]}
  ]
});

d1.save(function(err, result) {
  if (err) throw err;
  console.log();
  console.log(result.permissions);

  //var update = result; // works fine
  //var update = result.toObject(); // works fine
  //var update = _.clone(result._doc); // works fine
  var update = result._doc; // broken, mangles permissions array
  update.permissions = [
    {type: 'update', roles: [r1._id]},
    {type: 'create', roles: [r1._id]},
    {type: 'read', roles: [r1._id]}
  ];

  result.set(update);
  result.save(function(err, result) {
    if (err) throw err;
    console.log();
    console.log(result.permissions);
    mongoose.connection.close();
  });
});
```

Output (for broken case):
```
Permissions:
[ { type: 'create',
    roles: [ 55e0fd04535ee2fd64ebdd9e, 55e0fd04535ee2fd64ebdd9f ] },
  { type: 'read', roles: [ 55e0fd04535ee2fd64ebdd9e ] } ]

[{ type: 'create',
  roles: [ 55e0fd04535ee2fd64ebdd9e, 55e0fd04535ee2fd64ebdd9f ] }
{ type: 'read', roles: [ 55e0fd04535ee2fd64ebdd9e ] }]

Permissions:
[ { type: 'update', roles: [ 55e0fd04535ee2fd64ebdd9e ] },
  { type: 'create', roles: [ 55e0fd04535ee2fd64ebdd9e ] },
  { type: 'read', roles: [ 55e0fd04535ee2fd64ebdd9e ] } ]

[{ type: 'update', roles: [ 55e0fd04535ee2fd64ebdd9e ] }
{ type: 'update', roles: [ 55e0fd04535ee2fd64ebdd9e ] }
{ type: 'update', roles: [ 55e0fd04535ee2fd64ebdd9e ] }]
```

I believe this is unexpected output because the given key/value pairs are not fully mapped. The schema setter for the property is displaying the correct information, but only the first array element is actually being mapped (to every array element).

The difference between this test and my local issue is, the object I am using locally for `Document.set(<object>)` is not derived from the `Document._doc` object, however, I am getting the same results as seen above and I am trying to determine if my issue is related to this anomaly in any way.

Any clues as to why 1/4 options fails, and why it occurs when they all display the same information during the setter function call?

Thanks in advance!

**Comment by vkarpov15** [comment]
Manipulating `_doc` directly is not really supported... What are you trying to achieve by changing `_doc` directly?

**Comment by zackurben** [comment]
@vkarpov15 I am only manipulating `_doc` in this example because it shows the same strange output, I am experiencing. Have you ever seen such an anomaly, where the setter has the correct information, but still stores the incorrect information?

It seems like the first array element is marked as `modified`, but so is the entire array, so rather than iterate the elements, the array is just overwritten with the new one. In this case, its overwriting every element in the new array with the one that was marked as `modified`.

---
tl;dr: I am not using `_doc` in my code where I noticed this issue, but I am experiencing the same strange output. In my case, the update object isnt an instance of a mongoose document so I cant use `.toObject` nor does using `_.clone()` change the behavior.

**Comment by vkarpov15** [comment]
Because `_doc` is mongoose's backend storage of the raw JavaScript object. Shouldn't be too surprising that if you corrupt mongoose's internal state, bad things will start happening.

**Comment by zackurben** [comment]
@vkarpov15 so I looked into this more today, to get a better test case to show you.. turns out I found, and fixed, the issue in my local mongoose installation, and when I went to make a PR, my fix was already there...

It looks like @rosston beat me to the punch and I didn't realize I was behind a version (locally running 4.0.7 where I noticed the anomaly, and the fix was added in 4.0.8: https://github.com/Automattic/mongoose/commit/29c58b36c44d6264a52a0a134f491fdc002e785f).

Cheers and thanks for the lib!
